homeassistant/components/diagnostics/__init__.py

```diff
@@ -84,24 +97,49 @@ async def get(  # pylint: disable=no-self-use
         self, request: web.Request, d_type: str, d_id: str
     ) -> web.Response:
         """Download diagnostics."""
-        if d_type != "config_entry":
+        if d_type not in D_TYPES:
             return web.Response(status=404)
 
         hass = request.app["hass"]
-        config_entry = hass.config_entries.async_get_entry(d_id)
 
-        if config_entry is None:
-            return web.Response(status=404)
+        if d_type == D_TYPE_CONFIG_ENTRY:
+            config_entry = hass.config_entries.async_get_entry(d_id)
 
-        info = hass.data[DOMAIN].get(config_entry.domain)
+            if config_entry is None:
+                return web.Response(status=404)
 
-        if info is None:
-            return web.Response(status=404)
+            info = hass.data[DOMAIN].get(config_entry.domain)
 
-        if info["config_entry"] is None:
-            return web.Response(status=404)
+            if info is None or info[D_TYPE_CONFIG_ENTRY] is None:
+                return web.Response(status=404)
+
+            data = await info[D_TYPE_CONFIG_ENTRY](hass, config_entry)
+            filename = f"{config_entry.domain}-{config_entry.entry_id}"
+        else:
+            dev_reg = async_get(hass)
+            device = dev_reg.async_get(d_id)
+
+            if device is None:
+                return web.Response(status=404)
+
+            func_list = {
+                info[D_TYPE_DEVICE]
+                for config_entry in async_config_entries_for_device(hass, device)
+                if (info := hass.data[DOMAIN].get(config_entry.domain)) is not None
+            }
+
+            if not func_list:
+                return web.Response(status=404)
 
-        data = await info["config_entry"](hass, config_entry)
+            if len(func_list) == 1:
+                data = await next(func for func in func_list)(hass, device)
+            else:
+                data = {}
+                for config_entry in async_config_entries_for_device(hass, device):
+                    data[config_entry.domain] = await hass.data[DOMAIN][
```

Comment:
Can we guarantee that the same device can't be in two different config entries of the same domain?